proposal/color-4-new-spaces-js.d.ts.md

```diff
@@ -0,0 +1,540 @@
+  space?: ColorSpaceLCH
+): SassColor;
+
+changeChannels(
+  channels: {
+    [key in ChannelNameLAB]?: ChannelValue;
+  },
+  space?: ColorSpaceLAB
+): SassColor;
+```
+
+### `interpolate()`
+
+Returns a new SassColor with the result of mixing [`internal`] with `color2`.
+
+It accepts an optional float `weight`, which defaults to 0.5. Lower values will
+appear closer to [`internal`] and higher values will appear closer to `color2`.
+
+If `space` is set, interpolation will happen in that space. Otherwise it will
+happen in the color space for [`internal`].
+
+If `space` (or the color space of [`internal`] if no `space` argument is
+provided) is a PolarColorSpace (a color space with a polar angle `hue` channel),
+a `method` may be provided, which defaults to `shorter`.
+
+```ts
+interpolate(options: {
+  color2: SassColor;
+  weight?: number;
+  space?: RectangularColorSpace;
+}): SassColor;
+
+interpolate(options: {
+  color2: SassColor;
+  weight?: number;
+  space?: PolarColorSpace;
+  method?: HueInterpolationMethod;
+}): SassColor;
+```
+
+[`internal`]: ../spec/js-api/value/color.d.ts.md#internal
+
+### New Constructors
+
+Because the value of each channel may be a string, number, or `null`, this
+algorithm checks if an option with a key exists, instead of checking if it is
+set.
+
+* If `options.space` is not set, follow the legacy procedure for [construction].
+
+* Let `space` be a string with the value of `options.space`.
+
+* To calulate the channels:
+
+  * If `options.red` exists:
+
+    * Let `channel1` be the value of `options.red`.
+
+    * Let `channel2` be the value of `options.green`.
+
+    * Let `channel3` be the value of `options.blue`.
+
+  * If `options.saturation` exists:
+
+    * Let `channel1` be the value of `options.hue`.
+
+    * Let `channel2` be the value of `options.saturation`.
+
+    * Let `channel3` be the value of `options.lightness`.
+
+  * If `options.whiteness` exists:
+
+    * Let `channel1` be the value of `options.hue`.
+
+    * Let `channel2` be the value of `options.whiteness`.
+
+    * Let `channel3` be the value of `options.blackness`.
+
+  * If `options.x` exists:
+
+    * Let `channel1` be the value of `options.x`.
+
+    * Let `channel2` be the value of `options.y`.
+
+    * Let `channel3` be the value of `options.z`.
+
+  * If `options.chroma` exists:
+
+    * Let `channel1` be the value of `options.lightness`.
+
+    * Let `channel2` be the value of `options.chroma`.
+
+    * Let `channel3` be the value of `options.hue`.
+
+  * If `options.a` exists:
+
+    * Let `channel1` be the value of `options.lightness`.
+
+    * Let `channel2` be the value of `options.a`.
+
+    * Let `channel3` be the value of `options.b`.
+
+* If `options.alpha` is set:
+
+  * Let `alpha` be a Sass number with a value of `options.alpha`
+
+  * Set [`internal`] to the result of
+    [`color(channel1 channel2 channel3 / alpha)`]
+
+* Otherwise, set [`internal`] to the result of
+  [`color(channel1 channel2 channel3 )`]
```

Comment:
This seems to throw away the `space` value. Kust passing channel numbers to the `color()` function (e.g. `color(1 2 3)`) isn't valid. In addition, the CSS `color()` function _can't_ create a color with every space—colors that have their own functions, like `rgb()` or `lab()` _must_ be created with those functions.

It's also not clear to me why this procedure is defined here with a big conditional instead of having each constructor overload below define its own individual procedure.